### ISSUE TITLE
Display undocumented files and symbols

### ISSUE THREAD

**Issue by MrSquaare** [mosra/m.css#72]
Hello,

Is there a way to still display the files and functions of a library even if it is not documented?
Because I would like to have a more ergonomic documentation of the CSFML library (https://www.sfml-dev.org/download/csfml/index.php) but it is not fully documented

Thank you.
Sincerely

**Comment by mosra** [comment]
Hi!

Looking at the sources (e.g. [here](https://github.com/SFML/CSFML/blob/master/include/SFML/Network/Http.h)) I think it's quite close to being ideal :) The only thing you'd need in order to make the theme accept the files is adding a short `\file` statement at the top like below — otherwise, because it's a C library and there is no namespace or anything, the whole file gets ignored.

```cpp
#ifndef SFML_HTTP_H
#define SFML_HTTP_H

////////////////////////////////////////////////////////////
// \file
// \brief A HTTP client
////////////////////////////////////////////////////////////

…
```

Here I just used the brief description of the C++ `sf::Http` class as a file brief. I think this could be enough to make it recognize mostly everything.

Regarding your actual question, the theme was designed to throw away undocumented things since the beginning and the behavior is hardcoded on quite a lot of places. It should be definitely possible to implement that, but right now I'm extremely overwhelmed with other things, so I can't give you even an estimate when that could be. Nevertheless, I think if you'd do the above, you'd get quite far — it's not like the library would be absolutely documentation-less ;)

**Comment by MrSquaare** [comment]
Hello,

Thanks!

**Comment by mosra** [comment]
Okay, I'll keep this open to track the implementation for undocumented functions.

Let me know if the above suggestion worked out for you :)

**Comment by mosra** [comment]
For anybody who will be working on this in the future: Doxygen XML output currently contains everything, including private class members, anonymous namespaces and other cruft even if `EXTRACT_PRIVATE` and `EXTRACT_ANON_NSPACES` etc. is disabled (which is partially also the reason why current implementation was done the way it is). That needs to be fixed on Doxygen's side before we support for undocumented symbols can be implemented on m.css side.

Related issues (fixing this in Doxygen would close these):

- https://github.com/doxygen/doxygen/issues/2187
- https://github.com/doxygen/doxygen/issues/4504
- https://github.com/michaeljones/breathe/issues/399

**Comment by mosra** [comment]
@kobalicek So, in order to implement this, I'm thinking about taking some drastic shortcuts to avoid a maintenance nightmare -- for example preprocessing the XML to always put some static content in `<briefdescription>` in case both that and the `<detaileddescription>` is empty. Opinions on that? :)

Again, the major blocker is the above issue with *all stuff* being exposed by the Doxygen -- in order to make the output bearable, non-virtual undocumented private members, file-local members and contents of anonymous namespaces will still be ignored. And I'm not planning to support `EXTRACT_ANON_NSPACES`, `EXTRACT_LOCAL_CLASSES` etc. as I don't see a place for such things in end-user documentation.

However, the main thing I'm not sure about is the desired behavior -- as far as I'm aware, with the vanilla HTML output, a symbol that has no documentation whatsoever can't be linked to (it's not clickable in the output and all `@ref`s to it will produce a warning), doesn't appear in search and is generally hidden from everywhere except the single page that lists it. Is this what's wanted here? I'm not sure about usefulness of that behavior. The only option which I think could affect this is `EXTRACT_ALL`, but that one will also expose *everything else*, private, file-local members etc. -- neither that is desired. Or is there some other off-by-default option that affects this? I have no problem with behaving differently from the vanilla HTML output, but if a symbol can't be linked to or searched if it has no docs and there's no way to override that, then I'm afraid adding support for undocumented files/symbols on the theme side will not solve the problem fully and thus is not really useful.

**Comment by kobalicek** [comment]
I see it the following way:

  - If you use `\addtogroup XXX` followed by `\{` then I think everything until `\}` should be put into the documentation as you explicitly added the content into that group. However, if such undocumented symbol is not in any group then I'm not sure what to do, I would consider it not worth indexing. It could mean it's just internal not worth documenting. Maybe configurable through `HIDE_UNDOC_CLASSES`?
  - If you see undocumented class member in a documented class then I would assume that it should be visible unless you enable `HIDE_UNDOC_MEMBERS`.
  - Non-existing links to member and functions - If a class member is not documented and doesn't have it's own link I would link it to the "members" section in the class overview (basically where it appears).

I started writing my own documentation extractor to experiment with the XML doxygen generates so I started understanding what's extracted and what can you do with it. One interesting problem I discovered is that the XML doesn't provide all the information the HTML output provides. Doxygen removes some information before writing the XML so I can understand that some feature will never be possible (for example anonymous structs / unions in compounds are impossible to extract from the XML). This means that I don't expect much to be done especially if you follow the doxygen XML output.

**Comment by mosra** [comment]
> I would consider it not worth indexing.

Not sure about that. There are projects not using the grouping feature but grouping by header instead. AFAIK that's the default behavior of Doxygen -- even though it'll show undocumented classes and undocumented class members by default, free functions that are in an undocumented file will not appear in the output at all. That, to me, feels like an unnecessarily OOP-centric behavior and I think if showing undocumented symbols, then definitely these as well.

(See, that's one problem I have with showing undocumented stuff -- it's hard to draw a line between what *is* meant to be a part of the output and what not.)

> [...] configurable through `HIDE_UNDOC_CLASSES`
> [...] unless you enable `HIDE_UNDOC_MEMBERS`

Since I still believe undocumented symbols should not be part of the output by default, this behavior needs to be opt-in, not opt-out. Do you have any idea if there's a standard Doxygen option (that defaults to "not show") and could be reused for this? Both the `HIDE_*` options default to `OFF`, i.e. "show". For the time being, the theme would introduce a new, off-by-default option `M_SHOW_UNDOCUMENTED`.

> Non-existing links to member and functions - If a class member is not documented and doesn't have it's own link I would link it to the "members" section in the class overview

This was about behavior of Doxygen itself -- if it encounters and undocumented symbols, it will refuse to link to it. And there's nothing I can do on the theme side, the XML already has the links ~~broken~~ nonexistent. Again, in my eyes, this is the *main* problem which makes this feature rather useless :/

> I started writing my own documentation extractor to experiment with the XML doxygen generates

Friendly suggestion: **don't** ;) To be frank, knowing what I know now, I would not even *attempt* to try to parse the XML, but rather go straight to creating a full Doxygen replacement (which, well, is the only viable solution in the long term anyway). The issues are too many to list, and the inconsistency between HTML and XML output is the *least* of its problems (that one is actually easily fixable, unlike many others). To get an idea, see all the issues labelled <kbd>3rd party</kbd> (and the "3rd party TODO" column in the [Doxygen project](https://github.com/mosra/m.css/projects/5)). 

- - -

**Anyway**, so I'm not just blabbering around: there's a new branch `undocumented`, in particular commit df25c7c21858d855f5d63b5ff09d4701da747798, which adds a new `M_SHOW_UNDOCUMENTED` option. If you turn it `ON`, it'll include all non-private non-local symbols (files, directories, namespaces, classes, functions, typedefs, enums, defines, variables ...) it finds, and, for the time being, show them with a placeholder brief description. It's done in the simplest way possible in order to be maintainable. And since I won't personally be using such a feature, **feedback is very welcome**:

- Is the placeholder text okay? Should it say something else? Nothing at all? Should the text be configurable?
- Is it showing everything you want? Is it not showing too much? AFAIR, you should be able to configure Doxygen using various `IGNORE` patterns etc. to not include unwanted files/directories/namespaces in the docs, but what already gets into the XML and is not private, file-local or in an anonymous NS, will get exposed. I'm not planning to support the `HIDE_UNDOC_*` options, since they don't cover the complete spectrum (what if I want to hide only undocumented enums? or namespaces?) and it would be actually quite complex to support compared to how the feature is done now. Use the preprocessor, `IGNORE` options or `@cond` / `@endcond` to achieve the desired effect.
- Is search and general discovery of undocumented symbols good enough? Something missing?

**Comment by codecat** [comment]
The `undocumented` branch works well. As for feedback, I think the placeholder text (`No documentation present for this symbol.`) kinda shows up too much for me, which can be a little distracting:

![image](https://user-images.githubusercontent.com/136534/55151988-b906af00-514f-11e9-8cd8-b20a93f3e266.png)

So, not showing the text at all would probably be a good way to go about it. (Or at the very least, make it configurable, like you said.)

In my case, it is indeed showing everything I want because I am generating a header file specifically for documentation, so the header file has everything I need anyway. I also have no issues with symbol discovery or search. It works very well.

(Ps: I really like this theme! Very nice work! ❤️)

**Comment by mosra** [comment]
@codecat Thanks for the :heart: and the valuable feedback :)

Good to know it works for you -- I have no bigger "undocumented" project to test this on and so I kinda rely on external feedback here :) 

I thought about this a bit more and I think I'll go completely without the "No documentation present for this thing" text, since it's really just noise, yes. The enums need special handling to avoid showing them twice (yup, it's because of this option). I'll look into that and then merge all this to master.

**Comment by kobalicek** [comment]
TIP: `ENUM_VALUES_PER_LINE=0` in doxygen config controls that "I see enums twice" problem. I have done the same to only have one table to list enumerations. However, it has one caveat - you won't see the value of the enum anymore.

**Comment by codecat** [comment]
I think that is acceptable for me currently, thanks for the tip!

**Comment by mosra** [comment]
:shipit: 

The `undocumented` branch is now merged into master as 3044b35e325928d1c5c78ea1ec5ca912bfca56f0, [docs listing the caveats are here](https://mcss.mosra.cz/doxygen/#showing-undocumented-symbols-and-files). 

@codecat, in bf8dfed93d731ab0c188e06781a86886827c6f71 I did a bunch of special-casing to avoid showing undocumented enum values twice. If you could test and confirm / complain if it is working (or not), that would be great :)

**Comment by codecat** [comment]
Just tried the master branch, this looks good! Thanks! :D
